crates/symbolicator-service/src/services/symbolication/sourcemap.rs

```diff
@@ -0,0 +1,451 @@
+        }
+    },
+    {
+        dist: null,
+        name: '~/assets/bundle.min.js.map',
+        id: 42,
+        sha1: 023875,
+        headers: {}
+    },
+    {
+        dist: 'foo',
+        name: '~/assets/bundle.min.js',
+        id: 123,
+        sha1: 0238758,
+        headers: {}
+    }
+]
+
+// Match artifact from the list based on the `dist and `name`: (`null` and `~/assets/bundle.min.js`)
+{
+    dist: null,
+    name: '~/assets/bundle.min.js',
+    id: 1337,
+    headers: {
+        'X-SourceMap': 'bundle.min.js.map'
+    }
+}
+
+// Find sourcemap reference:
+    - either SourceMap header
+    - or X-SourceMap header
+    - or find pragma inside the file contents itself
+
+// Resolve sourcemap reference on top of abs_path of the frame we are processing
+abs_path: 'http://example.com/assets/bundle.min.js
+sm_ref: `bundle.min.js.map`
+=> 'http://example.com/assets/bundle.min.js.map`
+
+// Transform resolved sm_ref to use wildcards:
+=> '~/assets/bundle.min.js.map'
+
+// Match sourcemap artifact from the list based on the `dist` and resolved sourcemap ref: (`null` and '~/assets/bundle.min.js.map')
+{
+    dist: null,
+    name: '~/assets/bundle.min.js.map',
+    id: 42,
+    headers: {}
+}
+
+// Create SourceMapCache from the source and sourcemap files
+let smcache_writer = SourceMapCacheWriter::new(source_file, sourcemap_file).unwrap();
+let mut smcache_buf = vec![];
+smcache_writer.serialize(&mut smcache_buf).unwrap();
+
+// Cache the SourceMapCache using `sha1` of the abs_path artifact as the key `87324`
+
+// Process JS frame.
+
+*/
+
+pub type OwnedSourceMapCache = SelfCell<ByteView<'static>, SourceMapCache<'static>>;
+
+fn smcache_from_files(
+    source: &mut NamedTempFile,
+    sourcemap: &mut NamedTempFile,
+) -> OwnedSourceMapCache {
+    use std::io::Read;
+
+    let mut source_buf = String::new();
+    source.as_file_mut().read_to_string(&mut source_buf).ok();
+
+    let mut sourcemap_buf = String::new();
+    sourcemap
+        .as_file_mut()
+        .read_to_string(&mut sourcemap_buf)
+        .ok();
+
+    let smcache_writer = SourceMapCacheWriter::new(&source_buf, &sourcemap_buf).unwrap();
+    let mut smcache_buf = vec![];
+    smcache_writer.serialize(&mut smcache_buf).unwrap();
+
+    let byteview = ByteView::from_vec(smcache_buf);
+    SelfCell::try_new::<symbolic::sourcemapcache::SourceMapCacheError, _>(byteview, |data| unsafe {
+        SourceMapCache::parse(&*data)
+    })
+    .unwrap()
+}
+
+#[derive(Default, Debug)]
+struct FrameSourceMapping {
+    pub source_file: Option<NamedTempFile>,
+    pub source_artifact: Option<SearchArtifactResult>,
+
+    pub sourcemap_file: Option<NamedTempFile>,
+    pub sourcemap_artifact: Option<SearchArtifactResult>,
+
+    pub smcache: Option<OwnedSourceMapCache>,
+}
+
+impl SymbolicationActor {
+    async fn collect_stacktrace_artifacts(
+        &self,
+        request: &JsProcessingSymbolicateStacktraces,
+    ) -> HashMap<String, FrameSourceMapping> {
```

Comment:
As far as I can tell, the `FrameSourceMapping` struct is only used for `SourceMapCache` construction and never actually used outside of this function. This function might as well just return a map from `String` to `OwnedSourceMapCache`.
```suggestion
    async fn collect_stacktrace_artifacts(
        &self,
        request: &JsProcessingSymbolicateStacktraces,
    ) -> HashMap<String, OwnedSourceMapCache> {
```